Add heterodyne method to TimeSeries class

mattpitkin opened the pull request:
This adds a method similar to the TimeSeries demodulate method, however, this takes in the phase evolution within which to demodulate/heterodyne the array. By default this returns a double-sided complex output, rather than the single-sided value.

This PR also adds testing of the method.

Closes #1146 

cc @duncanmmacleod

pep8speaks commented:
Hello @mattpitkin! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`gwpy/timeseries/tests/test_timeseries.py`](https://github.com/gwpy/gwpy/blob/e08d6101317eac2e8c972600059e950e7ea8f71b/gwpy/timeseries/tests/test_timeseries.py):

> [Line 871:80](https://github.com/gwpy/gwpy/blob/e08d6101317eac2e8c972600059e950e7ea8f71b/gwpy/timeseries/tests/test_timeseries.py#L871): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

mattpitkin commented:
There seems to be some problem with the Python 3.6 build, but I don't think its specific to this PR. Maybe a Travis issue, or something in the build environment?

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1411,6 +1411,78 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
 
+    def heterodyne(self, phase, stride=1, exp=False, deg=True,
+                   singlesided=False):
+        """Compute the average magnitude and phase of this `TimeSeries`
```
@mattpitkin, if I'm reading your original feature request correctly, you want this method to always return a complex number rather than a `(mag, phase)` tuple, correct? So I think you can clean up the use case and docstring:

```python
    def heterodyne(self, phase, stride=1, singlesided=False):
        """Compute the average magnitude and phase of this `TimeSeries`
        once per stride after heterodyning with a given phase series

        Parameters
        ----------
        phase : `array_like`
            an array of phase measurements (radians) with which to heterodyne
            the signal

        stride : `float`, optional
            stride (seconds) between calculations, defaults to 1 second

        singlesided : `bool`, optional
            Boolean switch to return single-sided output (i.e., to multiply by 2
            so that the signal is distributed across positive frequencies only),
            default: False

        Returns
        -------
        out : `TimeSeries`
            magnitude and phase trends, represented as
            `mag * exp(1j*phase)` with `dt=stride`

        Notes
        -----
        This is similar to the :meth:`~gwpy.timeseries.TimeSeries.demodulate`
        method, but is more general in that it accepts a varying phase evolution,
        rather than a fixed frequency.

        Unlike :meth:`~gwpy.timeseries.TimeSeries.demodulate`, the complex
        output is double-sided by default, so is not multiplied by 2.

        See also
        --------
        TimeSeries.demodulate
            for a method to heterodyne at a fixed frequency
        """
        try:
            phaselen = len(phase)
        except Exception as exc:
            raise TypeError("Phase series is not array_like: {}".format(exc))
        if phaselen != len(self):
            raise ValueError(
                "Phase array must be the same length as the TimeSeries"
            )
        stridesamp = int(stride * self.sample_rate.value)
        nsteps = int(self.size // stridesamp)
        # stride through the TimeSeries and heterodyne
        out = type(self)(numpy.zeros(nsteps, dtype=complex))
        out.__array_finalize__(self)
        out.sample_rate = 1 / float(stride)
        phasearray = numpy.asarray(phase)  # make sure phase is a numpy array
        for step in range(nsteps):
            istart = int(stridesamp * step)
            iend = istart + stridesamp
            idx = numpy.arange(istart, iend)
            mixed = numpy.exp(-1j * phasearray[idx]) * self.value[idx]
            out.value[step] = 2 * mixed.mean() if singlesided else mixed.mean()
        return out
```

mattpitkin replied:
Thanks, your changes look good. I'd only ever be using it to output the complex time series rather than phases and magnitudes, and if anyone wants those quantities they are easy to generate.

mattpitkin replied:
I've made the changes based on your suggestion

alurban commented on gwpy/timeseries/tests/test_timeseries.py:
```diff
@@ -842,6 +842,53 @@ def test_demodulate(self):
         assert ph.unit == 'rad'
         utils.assert_allclose(ph.value, phase, rtol=1e-5)
 
+    def test_heterodyne(self):
```
This looks good, though if you use my other suggestion, the unit test can be cleaned up and simplified.

mattpitkin replied:
I'll implement your suggestion then clean up the test appropriately.

mattpitkin commented:
@alurban I'll look into adding a worked example. It shouldn't be too hard.

alurban commented on gwpy/timeseries/tests/test_timeseries.py:
```diff
@@ -842,6 +842,50 @@ def test_demodulate(self):
+        utils.assert_allclose(numpy.abs(het.value), amp, rtol=1e-4)
+        utils.assert_allclose(numpy.angle(het.value), phase, rtol=2e-4)
+
+        # test against demodulation for a fixed frequency
```
@mattpitkin, I think this block (lines 879 - 888) can safely be removed, since they're redundant with existing unit tests for `TimeSeries.demodulate`.

mattpitkin replied:
You suggestions look good. I've made the changes.

mattpitkin commented:
@alurban I've now added an example. Its quite involved and uses a CW hardware injection in S6 data. It also requires lalsuite being installed - is this ok, as I'm guessing this might break the docs build if it tries to run the example?

One line, which contains a long URL, breaks the PEP8 line length, but I've added a '# noqa` to stop it raising it as a problem. Is this ok?

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,184 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+
+        Examples
+        --------
+        Heterodyning can be useful in examining quasi-monochromatic signals
```
@mattpitkin, thanks for putting together this example, however as you pointed out earlier there are a couple of issues:

* The example is rather complicated, and
* The example uses routines from LAL that are beyond the scope of gwpy's documentation

Is it possible to split the difference and think of either a simpler use case that doesn't depend on LAL? For example, maybe try constructing a toy model for fdot/f of the 60 Hz U.S. power grid line, then heterodyne some O2 data with that model? Alternatively, you could just inject a pulsar signal (using `TimeSeries.inject`) into O2 data and then recover it with `TimeSeries.heterodyne`?

mattpitkin replied:
@alurban I've simplified the example to inject a sinusoid with a frequency that evolves with time. I then extract the signal's amplitude using `heterodyne` and compare it to demodulation at a fixed frequency.

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        >>>                     sample_rate=data.sample_rate, t0=data.t0)
+        >>> data = data.inject(signal)
+
+        We now bandpass the data around the signal frequency, and then
```
I propose a slight edit:

```python
        To recover the signal, we can bandpass the injected data around the
        signal frequency, then heterodyne using our phase model with a stride
        of 60 seconds:
```

mattpitkin replied:
Done.

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        >>> filtdata = data.bandpass(f0 - 0.5, f0 + 0.5)
+        >>> het = filtdata.heterodyne(phase, stride=60, singlesided=True)
+
+        Let's also demodulate the signal at a fixed freqeuncy for comparison
```
@mattpitkin, comparing to straight demodulation is slightly expensive for the docs and outside the scope of this example, so I propose we just get rid of `demod` and any references to it below.

mattpitkin replied:
Ok, done.

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        >>> plot = combined.plot()
+        >>> ax = plot.gca()
+        >>> ax.legend()
+        >>> ax.set_ylabel("Signal Strain Amplitude")
```
For consistency with other gwpy examples:

```suggestion
        >>> ax.set_ylabel("Strain amplitude")
```

mattpitkin replied:
Done.

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        >>> combined = TimeSeriesDict()
+        >>> combined['Heterodyned'] = het.crop(het.x0.value + 180).abs()
+        >>> combined['Demodulated'] = demod.crop(demod.x0.value + 180).abs()
+        >>> plot = combined.plot()
```
@mattpitkin, based on the comment above, the previous three lines involving `combined` can be deleted, and this line replaced with:

```python
        >>> plot = het.plot()
```

The `ax.legend()` line below can also be deleted.

mattpitkin replied:
Done.

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        inject such a signal and attempt to recover the amplitude.
+
+        >>> from gwpy.timeseries import TimeSeries, TimeSeriesDict
+        >>> data = TimeSeries.fetch_open_data('L1', 1131350417, 1131357617)
```
@mattpitkin, is there a reason so much data needs to be read in? Can you just read in the 180 seconds you need?

```suggestion
        >>> data = TimeSeries.fetch_open_data('L1', 1131350417, 1131350587)
```

mattpitkin replied:
It doesn't need to be quite so much (I just copied from the demodulation example), but it does need to be enough to show the difference between the heterodyning and demodulation, which requires more than 180 seconds.

mattpitkin replied:
I was working backwards through these, so have now seen you comment to remove the demodulation comparison.

mattpitkin replied:
I've not changed it just 180 seconds, but have halved the amount of data.

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
         """
+        # stride through the TimeSeries and heterodyne at a single frequency
+        phase = (2 * numpy.pi * f *
+                 self.dt.decompose().value *
```
@mattpitkin, after thinking about it, the following change is needed to prevent a memory leak due to double-precision floats:

```suggestion
                 self.dt.astype('float32').decompose().value *
```

mattpitkin commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+
+        Examples
+        --------
+        Heterodyning can be useful in examining quasi-monochromatic signals
```
Thanks. I've made this change.

mattpitkin commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        >>> fpeoch = 1131350417  # phase epoch
+        >>> amp = 1.5e-22  # signal amplitude
+        >>> phase0 = 0.4  # signal phase at the phase epoch
+        >>> import numpy
```
Done.

mattpitkin commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,143 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
+        frequency slowly decreases and is also Doppler modulated due to
+        the Earth's rotational and orbital motion.
+
+        We can show an example of demodulation by creating a simulated signal
```
Done.

alurban commented on gwpy/timeseries/tests/test_timeseries.py:
```diff
@@ -842,6 +842,40 @@ def test_demodulate(self):
         assert ph.unit == 'rad'
```
@mattpitkin, because of the switch to 32-bit floats, there is some loss in precision when calculating the phase of a demodulated signal. I think the loss is acceptable, but the unit tests for `TimeSeries.demodulate` will have to be updated.

With apologies for the extra hassle, can you make the following changes in `test_timeseries.py::TestTimeSeries::test_demodulate`?

* Line 830: change `rtol` to `1e-2`
* Line 838: change `rtol` to `1e-2`
* Line 843: change `rtol` to `1e-2`

Note, if you have the `pytest` package installed then you can run this unit test on the command-line to check that it behaves the way you want:

```bash
$ python -m pytest gwpy/timeseries/tests/test_timeseries.py::TestTimeSeries::test_demodulate
```

alurban commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1388,28 +1388,132 @@ def demodulate(self, f, stride=1, exp=False, deg=True):
         """
+        # stride through the TimeSeries and heterodyne at a single frequency
+        phase = (2 * numpy.pi * f *
+                 self.dt.astype('float32').decompose().value *
```
@mattpitkin, I was wrong before, apologies about that, can you remove the the conversion to `float32`?

```suggestion
                 self.dt.decompose().value *
```

**Context:** casting to `float32` causes a loss of precision, which breaks the unit tests for `TimeSeries.demodulate`. More to the point, the loss of precision is unacceptable since it leads to a systematic error in phase at the level of ~1 degree, which is on the same order as the current calibration uncertainty for advanced LIGO.

I was able to profile this method before and after your changes; before the change, demodulating 12 hours of 4 kHz data takes about 4.5 GB, while the same operation after your changes takes about 6.5 GB. I believe this is an acceptable trade-off to get the precision necessary to demodulate advanced LIGO's calibration lines, which is a mission-critical downstream use of this method.

alurban commented:
Thanks @mattpitkin, I think this PR is ready.

@duncanmmacleod, Matt and I have iterated a few times and I'm more than happy with the proposed changes, so I now pass the baton to you for final approval and merging.

duncanmmacleod commented:
@mattpitkin, I believe I have corrected the build failures in #1151, please rebase this PR against the upstream `master` branch.

mattpitkin commented:
@duncanmmacleod sure, it's been rebased with master now.

duncanmmacleod commented:
Thanks @mattpitkin
